blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
I have mixed feelings about this brick. First, What are the other use-cases besides Softmax, which can act axis-wide with minor code changes? Second, the assumption that all input arguments will need similar flattening treatment (and that it will work) seems limiting. Third, why wrap just the application method? This forces the instantiation of many bricks, as you do with 2 softmaxes above? Maybe it is better to go the full length - walk all methods of the wrapped brick, and wrap all applications?

Brief, I would add axes to softmax and not use WithExtraDims.